src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/HtmlCSharp/GoToImplementationHandler.cs

```diff
@@ -110,32 +117,57 @@ internal class GoToImplementationHandler : IRequestHandler<TextDocumentPositionP
             _logger.LogInformation($"Requesting {languageServerName} implementation for {projectionResult.Uri}.");
 
             var textBuffer = serverKind.GetTextBuffer(documentSnapshot);
-            var response = await _requestInvoker.ReinvokeRequestOnServerAsync<TextDocumentPositionParams, Location[]>(
+            var response = await _requestInvoker.ReinvokeRequestOnServerAsync<TextDocumentPositionParams, JToken>(
                 textBuffer,
                 Methods.TextDocumentImplementationName,
                 languageServerName,
                 textDocumentPositionParams,
                 cancellationToken).ConfigureAwait(false);
 
-            if (!ReinvocationResponseHelper.TryExtractResultOrLog(response, _logger, languageServerName, out var locations))
+            if (!ReinvocationResponseHelper.TryExtractResultOrLog(response, _logger, languageServerName, out var jToken))
             {
                 return null;
             }
 
-            if (locations.Length == 0)
+            cancellationToken.ThrowIfCancellationRequested();
+
+            // From some language servers we get VSInternalReferenceItem results, and from some we get Location results.
+            // We check for the _vs_id property, which is required in VSInternalReferenceItem, to know which is which.
+            if (jToken.Type == JTokenType.Array)
```

Comment:
This is a bit yuck, but I couldn't think of a better way. Essentially I want to return `VSInternalReferenceItem`s from Roslyn, so we need to support receiving them. The fact that we need to do anything here, and there isn't a converter that handles it, seems like a platform limitation for sure. Even if I did this the hard way, and changed Roslyn and Razor at the same time and did the dual insertion gubbins, it wouldn't help because we relay to multiple language servers, so the real problem this fixes is that our downstream servers might respond differently to the same method.

I couldn't think of a neat way to do this in the `LSPRequestInvoker`. The only thing I could think of was two generic params and return a tuple, for two possible types it could be, but then it would have to deserialize twice, which is not good. This has the advantage of being smart and actually inspecting the JSON, so we only pay the deserialization tax once. I guess that check could be provided by a predicate? It would essentially be `var (possibleResult1, possibleResult2) = _invoker.Request<Location[], VSInternalReferenceItem[]>(..., jtoken => jtoken.Has("_vs_id"), .., ...)`. Not very pretty.

Reply:
Hmm, yeah this does get a bit complicated. You can tell the spec was written with TS in mind with how casually they use union types. Is `VSInternalReferenceItem` a super-set of Location?

Reply:
No, that's the really frustrating bit. Location -> VSLocation -> VSInternalLocation are all fine, and have converters, but VSInternalReferenceItem is different. I could construct my own union type but keeping pace with changes would be worse than this IMO.

Reply:
If that's the case I think this is fine for now, but we should file an issue against Roslyn that whatever they return from this endpoint needs to meet (but may exceed) [the spec](https://microsoft.github.io/language-server-protocol/specification#textDocument_implementation).

Reply:
If we're going to do that (and this is exactly the sort of discussion I was hoping to have, so thank you) then this part of the change can be reverted, and we only need the change above where we stop losing information, and we create an issue on the platform to expand VSInternalLocation (or create a VSInternalReferenceLocationItem or similar) to include all of the capabilities of VSInternalReferenceItem but in a way that is a strict superset of Location.

Reply:
Yeah, if it's technically feasible I'd much rather make Roslyn adhere to the spec since I'm sure it will save us trouble in the future as well.

Reply:
Looks like (for now at least) we need to handle both types. There is a client capability that Roslyn can use to decide which one to use, but that doesn't help us on the server side.

So we either do this, or use `SumType<Location, VSInternalReferenceItem>` to handle it. Thoughts?

Reply:
Filed https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1470848 for the platform to follow up

Reply:
> So we either do this, or use SumType<Location, VSInternalReferenceItem> to handle it. Thoughts?

My vote is to take the SumType approach only because there's a precedence for it

Reply:
Will update to SumType